lib/mutgen.c

```diff
@@ -147,40 +147,17 @@ mutation_matrix_print_state(mutation_model_t *self, FILE *out)
 }
 
 static int MSP_WARN_UNUSED
-mutation_matrix_check_validity(mutation_matrix_t *self)
+probabilities_check_validity(size_t num_alleles, double *distribution)
 {
-    int ret = 0;
-    size_t j, k;
-    double prob, min_prob;
-    double *mutation_row;
-
+    double prob = 0;
+    double min_prob = 0;
     /* Check probabilities sum to one */
-    prob = 0;
-    min_prob = 0;
-    for (j = 0; j < self->num_alleles; j++) {
-        prob += self->root_distribution[j];
-        min_prob = GSL_MIN(min_prob, self->root_distribution[j]);
-    }
-    if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
-        ret = MSP_ERR_BAD_ROOT_PROBABILITIES;
-        goto out;
+    for (size_t j = 0; j < num_alleles; j++) {
+        prob += distribution[j];
+        min_prob = GSL_MIN(min_prob, distribution[j]);
     }
-    for (j = 0; j < self->num_alleles; j++) {
-        prob = 0;
-        min_prob = 0;
-        mutation_row = self->transition_matrix + j * self->num_alleles;
-        for (k = 0; k < self->num_alleles; k++) {
-            prob += mutation_row[k];
-            min_prob = GSL_MIN(min_prob, mutation_row[k]);
-        }
-        if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
-            ret = MSP_ERR_BAD_TRANSITION_MATRIX;
-            goto out;
-        }
-    }
-
-out:
-    return ret;
+    bool valid = (doubles_almost_equal(prob, 1.0, 1e-12) && min_prob >= 0.0);
+    return (valid ? 0 : MSP_ERR_BAD_PARAM_VALUE);
 }
```

Comment:
Haha, that's what I had before and @petrelharp suggested making the naming consistent with other functions. :) The way I've reconciled this in my mind is there are two kinds of functions: (A) error checking functions vs (B) just general purpose functions evaluating whether some condition is true or false. This function is kind of in a grey area between the two. In this recent commit I went for (A) but I think simpler is better so i think (B) is better choice.

I do think that _IF_ the return convention for this function is bool like a (B) and not error code like an (A) then the function name should suggest it is like an (B) function and sounds different than an (A) function (in other words intentionally inconsistent with naming of (A) functions because the return type is different in an extremely important way). For this reason I recommend prefixing such function names with `ok_....` or `is_....` etc... This makes is more clear to the reader of a function call that indeed it is the kind of function that returns a boolean rather than an int return code.

Is there any convention already in the codebase about function names that suggestion bool funtion return type?

If not, I can suggest `is_...` because the C++ library uses that a lot I believe, C library too probably.

Reply:
I've tended to do `ok_...` in my code, I'm not sure why. Maybe because `ok_` isn't jarring with plurals, unlike `is_` which is grating with plurals: e.g. `is_valid_proabilities`.
Regardless, with all these functions returning int values of `0` meaning "good/valid" I think it's good idea to make sure a call to a function returning a bool in which `zero`/`false` means "bad/invalid" is different.

Reply:
I don't think there's a convention for ``is_`` or ``ok_``. We could use ``are_valid_probablities``, if you like. It just seemed to read well to me in the intended usage ``if (!valid_probablities)`` reads literally "if not valid probablilities" to me. Since it's an internal static function that's only used in one place I'm not too bothered either way.